packages/map-toolkit/src/viewport/use-viewport-state.ts

```diff
@@ -0,0 +1,280 @@
+      const subscribers = componentSubscribers.get(instanceId);
+      if (subscribers) {
+        for (const onStoreChange of subscribers) {
+          onStoreChange();
+        }
+      }
+    }
+  });
+
+  busUnsubscribers.set(instanceId, unsub);
+}
+
+/**
+ * Cleans up the bus listener if no React subscribers remain.
+ *
+ * @param instanceId - The unique identifier for the viewport
+ */
+function cleanupBusListenerIfNeeded(instanceId: UniqueId): void {
+  const subscribers = componentSubscribers.get(instanceId);
+
+  if (!subscribers || subscribers.size === 0) {
+    // No more React subscribers - clean up bus listener
+    const unsub = busUnsubscribers.get(instanceId);
+    if (unsub) {
+      unsub();
+      busUnsubscribers.delete(instanceId);
+    }
+
+    // Clean up all state
+    viewportStore.delete(instanceId);
+    componentSubscribers.delete(instanceId);
+    subscriptionCache.delete(instanceId);
+    snapshotCache.delete(instanceId);
+  }
+}
+
+/**
+ * Creates or retrieves a cached subscription function for a given instanceId.
+ * Uses a fan-out pattern: 1 bus listener -> N React subscribers.
+ * Automatically cleans up viewport state when the last subscriber unsubscribes.
+ *
+ * @param instanceId - The unique identifier for the viewport
+ * @returns A subscription function for useSyncExternalStore
+ */
+function getOrCreateSubscription(
+  instanceId: UniqueId,
+): (onStoreChange: () => void) => () => void {
+  const subscription =
+    subscriptionCache.get(instanceId) ??
+    ((onStoreChange: () => void) => {
+      // Ensure single bus listener exists for this instanceId
+      ensureBusListener(instanceId);
+
+      // Add this React subscriber to the fan-out set
+      let subscribers = componentSubscribers.get(instanceId);
+      if (!subscribers) {
+        subscribers = new Set();
+        componentSubscribers.set(instanceId, subscribers);
+      }
+      subscribers.add(onStoreChange);
+
+      // Return cleanup function for THIS React subscriber
+      return () => {
+        const subscribers = componentSubscribers.get(instanceId);
+        if (subscribers) {
+          subscribers.delete(onStoreChange);
+        }
+
+        // Clean up bus listener if this was the last React subscriber
+        cleanupBusListenerIfNeeded(instanceId);
+      };
+    });
+
+  subscriptionCache.set(instanceId, subscription);
+
+  return subscription;
+}
+
+/**
+ * Creates or retrieves a cached snapshot function for a given instanceId.
+ * The object returned gets equality checked, so it needs to be stable or React re-renders unnecessarily.
+ *
+ * @param instanceId - The unique identifier for the viewport
+ * @returns A snapshot function for useSyncExternalStore
+ */
+function getOrCreateSnapshot(instanceId: UniqueId): () => MapViewportPayload {
+  const fallback = { ...defaultSnapshot, id: instanceId };
+  const snapshot =
+    snapshotCache.get(instanceId) ??
+    (() => viewportStore.get(instanceId) ?? fallback);
```

Comment:
to avoid creating a new object every time getOrCreateSnapshot is called, move this inside the snapshot variable default value:

```suggestion
  let fallback: MapViewportPayload | undefined;
  const snapshot =
    snapshotCache.get(instanceId) ??
    (() => {
      if (!fallback) {
        fallback = { ...defaultSnapshot, id: instanceId };
      }
      return viewportStore.get(instanceId) ?? fallback;
    });
```